### PR TITLE
P4C-DPDK - Add support for Hash extern

### PR DESCRIPTION
**Pull request by github-sajan**

### CODE REVIEW

File: backends/dpdk/dpdkHelpers.cpp [hunk 1]
```diff
@@ -159,10 +159,42 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
         auto mi = P4::MethodInstance::resolve(m, refmap, typemap);
         if (auto e = mi->to<P4::ExternMethod>()) {
             if (e->originalExternType->getName().name == "Hash") {
+                auto di = e->object->to<IR::Declaration_Instance>();
+                auto declArgs = di->arguments;
+                auto hash_alg = declArgs->at(0)->expression;
```
Comment (usha1830) [inline backends/dpdk/dpdkHelpers.cpp:169]: Please check the arg size prior to accessing.

**Review by usha1830 — APPROVED** [review]
One minor comment. Please address

File: backends/dpdk/dpdkHelpers.cpp [hunk 2]
```diff
@@ -159,10 +159,42 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
...
+                unsigned hashAlgValue = 0;
+                if (hash_alg->is<IR::Constant>())
+                    hashAlgValue = hash_alg->to<IR::Constant>()->asUnsigned();
+                cstring hashAlgName = "";
+                if (hashAlgValue == JHASH0 || hashAlgValue == JHASH5)
+                    hashAlgName = "jhash";
+                else if (hashAlgValue >= CRC1 && hashAlgValue <= CRC4)
+                    hashAlgName = "crc32";
+
+                IR::Vector<IR::Expression> components;
```
Comment (psivanup) [inline backends/dpdk/dpdkHelpers.cpp:179]: Limit the scope for "components", move the declaration to if block.

**Review by psivanup — APPROVED** [review]
Looks fine to me. Gave a "nice to have" comment.

File: backends/dpdk/dpdkHelpers.cpp (hunk 2 above)
Comment (psivanup) [reply backends/dpdk/dpdkHelpers.cpp:179]: not addressed?

File: backends/dpdk/dpdkHelpers.h [hunk 3]
```diff
@@ -178,6 +179,8 @@ class ConvertStatementToDpdk : public Inspector {
     cstring append_parser_name(const IR::P4Parser* p, cstring);
     void set_parser(const IR::P4Parser* p) { parser = p; }
     bool handleConstSwitch(const IR::SwitchStatement* a);
+    void processHashParams(const IR::Argument* field,
+                           IR::Vector<IR::Expression>& components);
```
Comment (psivanup) [inline backends/dpdk/dpdkHelpers.h]: functions not used outside class shall be declared private.

File: backends/dpdk/dpdkHelpers.cpp (hunk 2 above)
Comment (github-sajan) [reply backends/dpdk/dpdkHelpers.cpp:179]: "components" is now used in both if and else if part. So defining it outside if block

File: backends/dpdk/dpdkHelpers.cpp [hunk 4]
```diff
@@ -159,10 +159,81 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
...
+                IR::ListExpression *listExp = nullptr;
+
+                /* This process the two prototype for Hash Extern
```
Comment (usha1830) [inline backends/dpdk/dpdkHelpers.cpp]: This processes two prototypes of get_hash method of Hash extern

File: backends/dpdk/dpdkHelpers.h (hunk 3 above)
Comment (github-sajan) [reply backends/dpdk/dpdkHelpers.h]: Addressed

File: backends/dpdk/dpdkHelpers.cpp [hunk 5]
```diff
@@ -159,10 +159,80 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
...
+                    /// Constructor
+                    Hash(PSA_HashAlgorithm_t algo);
+
+                    1. Compute the hash for data.
+                        O get_hash<D>(in D data);
+
+                    2. Compute the hash for data, with modulo by max, then add base.
+                        O get_hash<T, D>(in T base, in D data, in T max);
+
+                        Here, DPDK targets allows constant expression for
```
Comment (usha1830) [inline backends/dpdk/dpdkHelpers.cpp]: add only to the comment here
DPDK targets allows "only" constant expression for

File: backends/dpdk/dpdkHelpers.cpp [hunk 6]
```diff
@@ -159,10 +159,80 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
...
+                    auto field   = (*e->expr->arguments)[1];
+                    auto max_val = (*e->expr->arguments)[2];
+
+                    /* Checks whether 'base' and 'max' param values are const values or not.
+                       If not, throw an error.
+                    */
+                    if (auto b = base->expression->to<IR::Expression>()) {
+                        if (!b->is<IR::Constant>())
+                            ::error("Expecting const expression '%1%' for 'base' value in"
+                            " Hash in DPDK Target", base);
```
Comment (usha1830) [inline backends/dpdk/dpdkHelpers.cpp]: You can use "get_hash method of Hash extern" instead of just "Hash"

File: backends/dpdk/dpdkHelpers.cpp [hunk 7]
```diff
@@ -159,10 +159,80 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
...
+                        if (b->is<IR::Constant>()) {
+                            maxValue = b->to<IR::Constant>()->asUnsigned();
+                            // Check whether max value is power of 2 or not
+                            if (maxValue == 0 || ((maxValue & (maxValue - 1)) != 0)) {
+                                ::error("Invalid Max value '%1%'. DPDK Target expect 'Max'"
+                                " value to be power of 2", maxValue);
+                            }
+                        } else {
+                            ::error("Expecting const expression '%1%' for 'Max' value in"
+                            " Hash in DPDK Target", max_val);
```
Comment (usha1830) [inline backends/dpdk/dpdkHelpers.cpp]: Likewise

File: backends/dpdk/dpdkHelpers.cpp [hunk 8]
```diff
@@ -346,6 +416,41 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
     return false;
 }
 
+/* This function process the hash parameters and stores it in
```
Comment (usha1830) [inline backends/dpdk/dpdkHelpers.cpp]: processes*

File: testdata/p4_16_samples_outputs/psa-hash-09.p4.spec [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+action NoAction args none {
+	return
+}
+
+action a1 args none {
+	mov m.Ingress_tmp h.ethernet.srcAddr
+	mov m.Ingress_tmp_0 h.ethernet.etherType
+	mov h.Ingress_tmp_1 h.ipv4
+	hash crc32 m.local_metadata_data  m.Ingress_tmp h.Ingress_tmp_1.dstAddr
```
Comment (usha1830) [inline testdata/p4_16_samples_outputs/psa-hash-09.p4.spec]: This looks wrong. The operands to hash instruction should come from same structure

File: testdata/p4_16_samples_outputs/psa-hash-08.p4.spec [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+header Ingress_tmp_0 instanceof ipv4_t
+
+action NoAction args none {
+	return
+}
+
+action a1 args none {
+	mov m.Ingress_tmp h.ethernet.srcAddr
+	mov h.Ingress_tmp_0 h.ipv4
+	hash crc32 m.local_metadata_data  m.Ingress_tmp h.Ingress_tmp_0.dstAddr
```
Comment (usha1830) [inline testdata/p4_16_samples_outputs/psa-hash-08.p4.spec]: Same here. The operands of hash instruction are coming from different structures which is wrong

**Review by usha1830 — CHANGES_REQUESTED** [review]
Changes look fine but some spec file output looks incorrect.
Please take a look.

**Review by usha1830 — APPROVED** [review]
Overall looks fine.
Let us handle the corner cases in separate PR.

**Review by psivanup — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**